File: 0067-add-binary/0067-add-binary.py
class Solution:
    def addBinary(self, a: str, b: str) -> str:
        def addBinaryStrings(a, b):
            result = ""
            carry = 0
            i = len(a) - 1
            j = len(b) - 1

            while i >= 0 or j >= 0 or carry:
                _sum = carry
                if i >= 0:
                    _sum += int(a[i])
                    i -= 1
                if j >= 0:
                    _sum += int(b[j])
                    j -= 1

                carry = _sum // 2
                result = str(_sum % 2) + result

            return result

        return addBinaryStrings(a, b)

File: 0067-add-binary/0067-add-binary.cpp
class Solution {
public:
    string addBinary(string a, string b) {
        int carry = 0;
        int m = a.size() - 1;
        int n = b.size() - 1;
        string result;

        while (m >= 0 || n >= 0 || carry) {
            int sum = carry;
            if (m >= 0) sum += a[m--] - '0';
            if (n >= 0) sum += b[n--] - '0'; 
            
            result.insert(result.begin(), sum % 2 + '0');
            carry = sum / 2; 
        }
        return result;
    }
};
